Generic Enumerable.Range method

AlexRadch opened the pull request:
Close #97689

AlexRadch committed:
Generic Enumerable.Range method

AlexRadch committed:
Add RangeCharTests

AlexRadch committed:
remove using

AlexRadch committed:
Delete build.txt

AlexRadch committed:
revert solution

AlexRadch committed:
Removed CreateTruncatingWithoutSign.

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Range.cs:
```diff
@@ -24,35 +24,63 @@ public static IEnumerable<int> Range(int start, int count)
+
+        public static IEnumerable<T> Range<T>(T start, int count) where T : IBinaryInteger<T>
+        {
+            if (count < 0)
```
Curious why this overload is using somewhat different validation flow compared to the existing method? Is it anticipating that `T.CreateTruncating` might be expensive?

eiriktsarpalis commented on src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs:
```diff
@@ -0,0 +1,1089 @@
+        //public static TheoryData<T, int> StartCountIncorrectData;
+
+        [Theory]
+        [MemberData("StartCountCorrectData")]
```
Can we use `nameof` here?

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Range.cs:
```diff
@@ -73,27 +101,33 @@ public override void Dispose()
                 _state = -1; // Don't reset current
             }
+
+            internal static int StartMaxCount(T start, T max)
```
I'm guessing this is effectively trying to measure a distance between `start` and `max`? I'm not sure if this is necessarily a well-defined operation for every kind of binary integer type (e.g. Gaussian integers)

cc @tannergooding

eiriktsarpalis commented on src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs:
```diff
@@ -0,0 +1,1089 @@
+        };
+    }
+
+    public class RangeBigIntegerTests : RangeIBinaryIntegerTests<BigInteger>
```
It would be interesting to see a few nonstandard numeric types that stress test the implementation, e.g. a modular arithmetic type. If for example I define a type representing all numbers modulo 7, what should be the behaviour of the new method when I do `Range(Modulo7.Zero, 8)`?

eiriktsarpalis reviewed (commented):
The build errors are caused by an [expected source breaking change](https://github.com/dotnet/runtime/issues/97689#issuecomment-2308916811) that this new API is introducing. The impacted call sites should be updated with an explicit `int` type parameter.

AlexRadch committed:
Added explicit (int) cast to Enumerable.Range.
It is expected source breaking change that this new API is introducing.

AlexRadch committed:
retrigger checks

eiriktsarpalis reviewed (changes requested):
Marking as changes requested since there are a few unresolved questions on a number of corner cases.

jkotas commented on src/libraries/Common/tests/System/IO/VirtualDriveHelper.Windows.cs:
```diff
@@ -87,7 +87,7 @@ char GetNextAvailableDriveLetter()
 
                 // A,B are reserved, C is usually reserved
-                IEnumerable<int> range = Enumerable.Range('D', 'Z' - 'D');
+                IEnumerable<int> range = Enumerable.Range((int)'D', 'Z' - 'D');
```
```suggestion
                IEnumerable<int> range = Enumerable.Range<int>('D', 'Z' - 'D');
```

jkotas replied:
Similar in other places. Explicit selecting the result type and leave it to the compiler to convert the arguments to it implicitly is less bug prone.

stephentoub commented:
Now that https://github.com/dotnet/runtime/pull/111685 has been merged, we should augment this PR to add the corresponding generic Range method to AsyncEnumerable as well. In general, we want to keep the APIs consistent across Enumerable and AsyncEnumerable.

AlexRadch committed:
Merge branch 'main' of https://github.com/dotnet/runtime into Linq_Range_IBinaryInteger

AlexRadch committed:
Refactor Enumerable.Range usage for type safety

Updated usage of Enumerable.Range to specify <int> type
parameter in VirtualDriveHelper.Windows.cs,
ReadOnlySequenceFactory.char.cs, and ArrayTests.cs.
This change improves code clarity and enhances type safety
in drive letter generation and ASCII range handling,
as well as strengthens assertions in unit tests.

AlexRadch committed:
Add generic AsyncEnumerable.Range<T> method for IBinaryInteger types

Introduces a new method `Range<T>(T start, int count)` in the `AsyncEnumerable` class for generating sequences of integral numbers for types implementing the `IBinaryInteger<T>` interface. This method is conditionally compiled for .NET 7.0 or greater and includes validation for the `count` parameter.

Updates `RangeTests` to cover various input scenarios for the new method, including edge cases for different integral types like `byte` and `long`. Additionally, enhances documentation for the existing `Range(int start, int count)` method.

AlexRadch commented:
> Now that #111685 has been merged, we should augment this PR to add the corresponding generic Range method to AsyncEnumerable as well. In general, we want to keep the APIs consistent across Enumerable and AsyncEnumerable.

@stephentoub Done!

stephentoub commented on src/libraries/System.Linq.AsyncEnumerable/ref/System.Linq.AsyncEnumerable.cs:
```diff
@@ -110,6 +110,9 @@ public static partial class AsyncEnumerable
         public static System.Collections.Generic.IAsyncEnumerable<int> Range(int start, int count) { throw null; }
+#if NET7_0_OR_GREATER
+        public static System.Collections.Generic.IAsyncEnumerable<T> Range<T>(T start, int count) where T : System.Numerics.IBinaryInteger<T> { throw null; }
+#endif
```
Thanks. This should be in its own System.Linq.AsyncEnumerable.netcore.cs file rather than ifdef'd here.

stephentoub commented on src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs:
```diff
@@ -35,5 +36,51 @@ static async IAsyncEnumerable<int> Impl(int start, int count)
+                for (int i = 0; i < count; i++, start++)
+                {
+                    yield return start;
+                }
```
@tannergooding, is this the right implementation? The above non-generic does:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + i;
}
```
Is this proposed implementation better or worse than:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + T.CreateTruncating(i);
}
```
?  I'm not sure what requirements we place on IBinaryInteger, in terms of guarantees we make about ++ actually incrementing the value (as opposed to, say, a float + 1 which might not actually increment).

stephentoub commented on src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs:
```diff
@@ -35,5 +36,51 @@ static async IAsyncEnumerable<int> Impl(int start, int count)
+                if (count < 0)
+                    count = int.CreateTruncating(max) - int.CreateTruncating(start);
+                return count;
+            }
```
@tannergooding, is this how we'd recommend doing overflow checking?

AlexRadch committed:
Add Range method for IAsyncEnumerable<T> in .NET Core

This commit introduces a new implementation of the `Range` method for `IAsyncEnumerable<T>` that supports `IBinaryInteger<T>` types, specifically for .NET Core. The method generates a sequence of integral numbers within a specified range, with input validation and exception handling for out-of-range values. Project files have been updated for conditional compilation targeting .NET Core, and the previous `Range` method for non-`IBinaryInteger<T>` types has been removed. Tests have also been updated to reflect these changes.

AlexRadch committed:
Add BigInteger support to RangeTests

- Introduced `System.Numerics` for `BigInteger` usage.
- Added test for `ArgumentOutOfRangeException` with `BigInteger` in `InvalidInputs_Throws`.
- Implemented tests for various `BigInteger` values in `VariousValues_MatchesEnumerable`.

AlexRadch committed:
Update RangeTests to include BigInteger values

Modified `RangeTests.cs` to test a wider range of `BigInteger` values instead of just `long`. Added a new test case for `AsyncEnumerable.Range<BigInteger>` that checks the behavior when skipping and taking elements from a large range. All tests executed successfully with no errors or failures.

AlexRadch committed:
Update InvalidInputs_Throws tests for AsyncEnumerable

Modified test cases in `InvalidInputs_Throws` in `RangeTests.cs` to validate `ArgumentOutOfRangeException` for negative count values in `AsyncEnumerable.Range` for `byte`, `long`, and `BigInteger` types. All tests passed successfully.

AlexRadch committed:
Fix assertions in ArrayTests to cast chars to ints

Updated assertions in `Array_ByValueOut` and `Array_ByValueOut_This` methods to cast characters in `testArray` to integers. This ensures proper comparison with the expected output of `Enumerable.Range<int>`, enhancing test accuracy.
